File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cs
public class Solution {
    public int Search(int[] nums, int target) {
        int N=nums.Length;
        int left=0;
        int right=N-1;
        while(left<=right)
        {
            int mid=(left+right)/2;
            if(target==nums[mid])  return mid;
            else if(nums[0]<=nums[mid])//if left portion is sorted ascending
            {
                if(target<nums[mid] && target>=nums[0])//make sure target in that sorted space only
                    right=mid-1;
                else
                    left=mid+1;
            }
            else//right portion is sorted ascending
            {
                if(target>nums[mid] && target<=nums[N-1])//make sure target in that sorted space only
                    left=mid+1;
                else
                    right=mid-1;
            }
        }
        return -1;
    }
}

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int N=nums.length;
        int start=0,end=N-1;
        while(start<=end)
        {
            int mid=start+(end-start)/2;
            if(nums[mid]==target)
            {
                return mid;
            }
            else if(nums[0]>nums[mid])//sorted range in right
            {
                if(nums[mid]<target && nums[N-1]>=target)//if our item in that range
                    start=mid+1;
                else
                    end=mid-1;
            }
            else//sorted range in left
            {
                if(nums[mid]>target && nums[0]<=target)//if our item in that range
                    end=mid-1; 
                else
                    start=mid+1;
            }
        }
        return -1;
    }
   
}